### PR TITLE
Add issue and pull request templates

### PR DESCRIPTION
**Pull request by lildude**
This PR adds an issue and pull request template to help peeps help themselves, and us.

Issue and PR templates have been a feature of GitHub.com for quite some time and I think it's about time we started using them too, so tadaaa 🎉

My aim with these initial templates is to guide peeps logging issues to try and help themselves and work out the cause before logging an issue.

For pull requests, I'm trying to ensure all the pre-reqs are met without us having to ask each time.

What do you think peeps?

### CODE REVIEW

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+## Description
+<!--- Describe your changes in detail -->
+
+## Checklist:
+<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
+<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
+- [ ] **I am adding a new extension to an existing language.**
+  - [ ] The new extension is used in hundreds of repositories on GitHub.com
+    - Search results for each extension:
+      <!-- Replace FOOBAR with the new extension. Repeat for each extension added. -->
```
Comment (pchaigno) [inline .github/PULL_REQUEST_TEMPLATE.md]: We should maybe encourage contributors to add keywords in the search query at this point, if the results contain files from other languages...?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+      - https://github.com/search?utf8=%E2%9C%93&q=extension%3AFOOBAR+NOT+nothack&type=Code&ref=searchresults
+  - [ ] I have included a real-world usage sample for all extensions added in this PR:
+    - Sample source(s):
+      - [URL to each sample source, if applicable]
+    - Sample license(s):
+  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.
```
Comment (pchaigno) [inline .github/PULL_REQUEST_TEMPLATE.md]: Given that the above points are mandatory, this one feels mandatory too. Do we want to make adding heuristic rules mandatory, or should we discuss this with contributors on a case-by-case basis?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+      -  https://github.com/search?utf8=%E2%9C%93&q=extension%3AFOOBAR+NOT+nothack&type=Code&ref=searchresults
+  - [ ] I have included a real-world usage sample for all extensions added in this PR:
+    - Sample source(s):
+      - [URL to each sample source, if applicable]
+    - Sample license(s):
+  - [ ] I have included a syntax highlighting grammar.
+  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.
+
+- [ ] **I am fixing a misclassified language**
+  - [ ] I have included a sample for all extensions added in this PR:
```
Comment (pchaigno) [inline .github/PULL_REQUEST_TEMPLATE.md]: Should this be `I have included a new sample for the misclassified language`?

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    - Sample source(s):
+      - [URL to each sample source, if applicable]
+    - Sample license(s):
+  - [ ] I have included a change to the heuristics to distinguish my language from others using the same extension.
+
+- [ ] **I am changing the source of a syntax highlighting grammar**
```
Comment (pchaigno) [inline .github/PULL_REQUEST_TEMPLATE.md]: I think we should require Lightshow links to highlight the improvements here.

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 2 above)
Comment (lildude) [reply .github/PULL_REQUEST_TEMPLATE.md]: I think we should keep it optional as we don't always need a heuristic and suggest it's only optional in CONTRIBUTING.md

File: .github/PULL_REQUEST_TEMPLATE.md (hunk 4 above)
Comment (lildude) [reply .github/PULL_REQUEST_TEMPLATE.md]: Good idea.

Comment (lildude) [reply .github/PULL_REQUEST_TEMPLATE.md]: How's that 👇 @pchaigno ?

Comment (pchaigno) [reply .github/PULL_REQUEST_TEMPLATE.md]: I was thinking of asking for two links, one for before, one for after, to make sure contributors highlight what we might otherwise overlook.
Sorry, I should have been clearer.

**Comment by pchaigno** [comment]
I'm thinking of this only now, but perhaps we should have an entry or several for contributors adding new items to the vendor, documentation, and generated lists?

**Comment by pchaigno** [comment]
Also, could you take a look when you've got some time @Alhadis? You've got a better eye than I for this sort of things.

**Comment by lildude** [comment]
@Alhadis and thoughts before I merge this?

File: .github/ISSUE_TEMPLATE.md [hunk 5]
```diff
@@ -0,0 +1,23 @@
+<!--- Provide a general summary of the issue in the Title above -->
+
+## Preliminary Steps
+
+Please review these preliminary steps before logging your issue. You may find the information referenced may answer or explain the behaviour you are seeing. It'll help us to know you've reviewed this information.
```
Comment (Alhadis) [inline .github/ISSUE_TEMPLATE.md]: People have a tendency to gloss over these things, so it might help to move this below the checklist. The list does open with *"Please confirm you have..."* so it opens up nice and direct. One immediately sees that something is required first.

I'd also wrap this paragraph in an HTML comment too, so lazy reporters don't give us messier reports. 😉

File: .github/ISSUE_TEMPLATE.md [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+- [ ] reviewed [How Linguist Works](https://github.com/github/linguist#how-linguist-works),
+- [ ] reviewed the [Troubleshooting](https://github.com/github/linguist#troubleshooting) docs,
+- [ ] considered implementing an [override](https://github.com/github/linguist#overrides),
+- [ ] verified an issue has not already been logged for your issue ([linguist issues](https://github.com/issues?utf8=%E2%9C%93&q=is%3Aissue+repo%3Agithub/linguist)).
+
+## Problem Description
+
+<!--- Provide a more detailed introduction to the issue itself, and why you consider it to be a bug -->
+
+**Affected repository:** <!-- URL to repository -->
```
Comment (Alhadis) [inline .github/ISSUE_TEMPLATE.md]: I'd make these tertiary subheadings, and reword the first one so there's no need for a trailing HTML comment:

~~~markdown
### URL of the affected repository

### Last modified on:
<!-- YYYY-MM-DD -->

... etc
~~~

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 7]
```diff
@@ -0,0 +1,46 @@
+<!--- Provide a general summary of your changes in the Title above -->
```
Comment (Alhadis) [inline .github/PULL_REQUEST_TEMPLATE.md]: I'd shorten this to *"Briefly describe what you're changing"*.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+
+## Description
+<!--- Describe your changes in detail -->
```
Comment (Alhadis) [inline .github/PULL_REQUEST_TEMPLATE.md]: Some PRs are simple to a point of being self-explanatory, so perhaps reword this to become:

> If necessary, go into depth of what this pull-request is doing.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+## Checklist:
+<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
+<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
+- [ ] **I am adding a new extension to an existing language.**
+  - [ ] The new extension is used in hundreds of repositories on GitHub.com
```
Comment (Alhadis) [inline .github/PULL_REQUEST_TEMPLATE.md]: The word "new" can be axed, because it's technically impossible to add an old extension.

I might reword this to *"I am associating a language with a new file extension"*, since it's clearer English.

File: .github/PULL_REQUEST_TEMPLATE.md [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    - Search results for each extension:
+      <!-- Replace FOOBAR with the new extension, and KEYWORDS with keywords unique to the language. Repeat for each extension added. -->
+      - https://github.com/search?utf8=%E2%9C%93&q=extension%3AFOOBAR+KEYWORDS+NOT+nothack&type=Code&ref=searchresults
```
Comment (Alhadis) [inline .github/PULL_REQUEST_TEMPLATE.md]: It would help to move the `FOOBAR+KEYWORDS` to the end of the URL, for easier copy+pasta. Although I find it amazing there's no front-end mechanism for running this search, yet there's a list of language colours buried away in https://github.github.com/linguist/ 😣

**Comment by Alhadis** [comment]
Holy crap, I'm so sorry for not getting back sooner. 😓 I've only been half-here so my Github activity has been very spotty.

**Comment by Alhadis** [comment]
(I've also been stuck using mobile, and GitHub isn't very friendly for mobile users... case-in-point, I'm adding a second comment because there's no button to edit the first 😂👌)

**Comment by pchaigno** [comment]
Wouhou! That should help reduce the time spent on each contribution :-)
